x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/execution/sample/SampleIterator.java

```diff
@@ -0,0 +1,327 @@
+                    advance(listener);
+                } else {
+                    if (stack.size() > 0) {
+                        nextPage(listener, stack.pop());
+                    } else {
+                        payload(listener);
+                    }
+                }
+            }, listener::onFailure));
+        } else if (currentStage > maxStages) {
+            throw new EqlIllegalArgumentException("Unexpected stage [{}], max stages in this sample [{}]", currentStage, maxStages);
+        } else {
+            finalStage(listener);
+        }
+    }
+
+    /*
+     * Creates a _msearch request containing maxStages (number of filters in the query) * number_of_join_keys queries.
+     * For a query with three filters
+     *          sample by host [any where uptime > 0] by os [any where port > 5] by op_sys [any where bool == true] by os
+     * and one pair of join keys values (host = H1, os = win10) the msearch will look like
+     * ...{"bool":{"must":[{"term":{"host":"H1"}},{"term":{"os":"win10"}},{"range":{"uptime":{"gt":0}}}]}},"terminate_after":3,"size":3}
+     * ...{"bool":{"must":[{"term":{"host":"H1"}},{"term":{"op_sys":"win10"}},{"range":{"port":{"gt":5}}}]}},"terminate_after": 3,"size":3}
+     * ...{"bool":{"must":[{"term":{"host":"H1"}},{"term":{"os":"win10"}},{"term":{"bool":true}}]}},"terminate_after": 3,"size":3}
+     */
+    private void finalStage(ActionListener<Payload> listener) {
+        log.trace("Final stage...");
+        // query everything and build sequences
+        // then remove the previous step from the stack, since we are done with it
+        Page page = stack.pop();
+        // for each criteria and for each matching pair of join keys, create one query
+        List<SearchRequest> searches = new ArrayList<>(maxStages * page.hits.size());
+
+        // should this one be a Set instead? Meaning, unique SequenceKeys? The algorithm should already be generating unique keys...
+        List<SequenceKey> sampleKeys = new ArrayList<>();
+        // get all composite key values as a list of list
+        List<List<Object>> allCompositeKeyValues = criteria.get(maxStages - 1).keyValues(page.hits);
+        for (List<Object> compositeKeyValues : allCompositeKeyValues) {
+            // take each filter query and add to it a filter by join keys with the corresponding values
+            for (SampleCriterion criterion : criteria) {
+                SampleQueryRequest r = criterion.finalQuery();
+                r.singleKeyPair(compositeKeyValues, maxStages);
+                searches.add(prepareRequest(r.searchSource(), false, EMPTY_ARRAY));
+            }
+            sampleKeys.add(new SequenceKey(compositeKeyValues.toArray()));
+        }
+
+        int initialSize = samples.size();
+        client.multiQuery(searches, ActionListener.wrap(r -> {
+            List<List<SearchHit>> finalSamples = new ArrayList<>();
+            List<List<SearchHit>> sample = new ArrayList<>(maxStages);
```

Comment:
Probably we can use `List<Set<SearchHit>>` here. It would make some operations faster (ie. `contains()`) and it would make the intention more clear.

Reply:
Isn't the order of requests kept in the first List...? The Set contains the results for a single query filter, that should not be so relevant as long as you can decide which filter you run first

Reply:
You're right. That order shouldn't matter.